platform_plugin_aspects/sinks/course_overview_sink.py

```diff
@@ -58,7 +86,22 @@ def serialize_item(self, item, many=False, initial=None):
         detached_xblock_types = get_detached_xblock_types()
 
         location_to_node = {}
-        items = modulestore.get_items(course_key)
+
+        # This call gets the entire course tree in order, because the
+        # get_items call does not guarantee ordering. It does not return
+        # detached blocks, so we gather them separately below.
+        course_block = modulestore.get_course(
+            course_key, revision="rev-opt-published-only"
+        )
+        items = self.get_xblocks_recursive(course_block)
```

Comment:
Pulling all of a course's blocks into memory like this might not be a good idea..

Holding just the parts of the blocks that we need (usage key, display_name, ...) would be more memory efficient than holding the whole block(s).

Reply:
I don't think this is related to this PR, but I'm seeing Course Overview and XBlock sink events in the logs when I just add a section or subsection to a course, but I haven't hit Publish yet. Are these structure blocks are automatically "published"?

I mention it because it seems to block the "new section/subsection" request response in Studio, which was already noticeably slow.

Reply:
Sure, I can refactor it to do some kind of generator thing to avoid holding everything in memory.

If this is getting triggered that should mean that `COURSE_PUBLISHED` for fired, but I'd guess the slowness you're seeing is in dev where all of the heavyweight Celery tasks are happening in-process. I wouldn't expect that slowness to be a thing that happens in a production setting. It's sub-second on the demo course for me.

Reply:
@pomegranited I took a stab at the memory consumption in the latest commit, but I'm not sure how much it helps. Depends on modulestore internals that I don't fully understand yet.

Reply:
Excellent -- serializing the blocks down to minimal properties should help, thank you!

>  but I'd guess the slowness you're seeing is in dev where all of the heavyweight Celery tasks are happening in-process

Oh absolutely, my poor little devstack isn't representative, just a good canary.